### DISCUSSION TITLE
Next major priority survey - IK or Performance?

### DISCUSSION THREAD

**Discussion by Frooxius** [Yellow-Dog-Man/Resonite-Issues#2066] [General]
Hello everyone!

## What's this about?

We're hosting a survey to help us decide the next major step for Resonite - if we should work on IK first or focus on major performance improvements. We'd like input from you - our community to help us make the final decision.

## Read the survey first
The survey contains details about what would both of these priorities involve and number of questions that will help us decide.

[View the survey here.](https://files.yellowdogman.com/apps/forms/embed/z8YB9CNpEZYSaTxDBm6i6zEf)

## Before you fill it out - discuss it here

We'd like to give you time to think about the survey, so you don't have to decide immediately - some things are better to sleep on for a bit - and don't worry, we have a bunch of other small things to tackle in the meanwhile.

We'd like this discussion to provide a way to ask us questions about these priorities and anything you might need more clarity on before you answer.

**Comment by SB15SKG** [comment]
Only big question I can think of:
Will whichever isn't picked be high on the priority list after the chosen is considered "complete"?

**Reply by Frooxius** [reply]
I'm sorry, but I don't understand your question. Would you be able to rephrase it please?

**Reply by epicEaston197** [reply]
They're probably asking if for example we chose performance if after that's done the IK rework would be next?

**Reply by SB15SKG** [reply]
As in, let's say Performance wins, will the IK "goal" be worked on in the future or vice-versa? I'd like for both to be implemented, but of course one needs to be prioritized over the other.

**Reply by Frooxius** [reply]
Yes. Both will be worked on at some point, this is more about which one gets to be focused on next.

In other words, picking one doesn't cancel the other one. Some of these, like performance, are also big ongoing projects, so they will be returned to regularly as the platform continues to develop.

**Reply by netcore10** [reply]
Honestly the earlier you can do performance, the better it is later.

**Comment by shininghero** [comment]
Intriguing, are the two choices that mutually exclusive that solving one won't help the other?  Usually my performance problems are due to VRIK and the data model making me CPU bound.

**Reply by Frooxius** [reply]
Yes, they're separate things. We can only work on one major thing at the time - so one of these has to be done first. I recommend reading the text of the survey, it goes into detail on this a bit more.

**Reply by shininghero** [reply]
Yeah, in hindsight there's a couple of other pain points in performance that I taught myself to ignore.  Like the occasional session hitching even in solo worlds.

If you're predicting double performance gains from this... Well, now I'm getting fired up myself.  I want to see that.

**Reply by Frooxius** [reply]
It's the conservative goal. Ideally we'd like to have higher increases, but it's hard to make solid predictions on stuff like this.

It will also depend on what's happening. Like some things will improve significantly, while some might not as much.

But the overall goal is to make the performance with regular use feel a lot better.

**Reply by LexiBasilisk** [reply]
To clarify, @shininghero - IK in this case wouldn't cover the data model rework- that would be covered under #702 which is a performance item.

As it mentions in this issue:
> However currently this comes with certain costs, as some common usage patterns end up causing significant amount of overhead. Notably any behaviors that are continuously changing the data model (such as IK or Dynamic Bones) have their changes explicitly tracked every single frame.

So the IK itself is not the performance issue, but rather the amount of changes it makes to the data model every frame. It's a symptom- rather than the underlying issue.

**Comment by Inucune2** [comment]
Is it possible to have the IK use a priority system (user configured) for updates?
EX: my hands/head/body updates take priority over hair/accessories/ancillary, and this would respect other user's priority settings?
A floor value of 0 (possibly reserved for the base IK), a default of 1 or 2, and probably a max of 16 or 32 should be high enough.
This could provide a simple honor-system method for people with many items to delay updating certain items when the IK gets overloaded.

**Reply by Frooxius** [reply]
I'm not really sure what this means in context of Resonite. IK doesn't have any priority system like that at the moment.

**Reply by LexiBasilisk** [reply]
> I'm not really sure what this means in context of Resonite. IK doesn't have any priority system like that at the moment.

My interpretation of this comment was that @Inucune2 was wondering about component rate limiting, e.g. setting things like DynamicBoneChains, and IK to update less frequently based on certain conditions, e.g. performance/framerate target, and prioritizing updates from specific users/specific instances of components over others.

So I believe @Inucune2 was wondering if culling/rate limiting the amount of IK/dynamic bone updates other users produce would be in the scope of the IK rework.

If I am incorrect in this interpretation, please feel free to clarify further, @Inucune2.

**Comment by epicEaston197** [comment]
My take away from this is we need better performance there's no use in having good full body tracking when you won't have the frames to experience it I've been reading negative steam reviews for a while now and a few mentioned the performance being a reason why they got droven away

not only that but clubs that want to perform in Resonite can't due to the limited performance you can only have 20 people in a session before your frame rate starts to tank and that's nothing compared to the clubs in VRC so if we even remotely want to have a chance of developing a club scene we need better performance

I've also experienced a club and Resonite...
The performance was not good

**Reply by epicEaston197** [reply]
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/76523692/17b06a06-433b-4364-93d0-f9ca29e3dacc)
this is all that's in the world
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/76523692/d69e5f00-fd42-411f-975f-49cc555e584c)

**Reply by charlie-sans** [reply]
i think kokolive is a bit harsh on preformance.

i might check out the code again when i work on the new kokolive version

**Comment by ModernBalloonie** [comment]
This is a hard one.

IK is something I definitely struggle with and I have a lot of friends who refuse to use Resonite due to this.

As a multi-platform avatar creator, an IK improvement would really help, as I struggle with it a lot. There are a lot of strange things i've had to work-around, and certain things that work so great in other platforms don't work as good in Resonite. The neck scrunching is super annoying, and I've had to calibrate my avatars obsessively to get it to work as well as it does for me.

Performance is super important, don't get me wrong, but the IK has been a source of constant struggling for me (and many friends who i've tried to bring in). Though, of course that's just my opinion.

Though I will say, I am used to the low performance, so my opinion might be a little bit biased.

**Reply by epicEaston197** [reply]
Ohh yeah I see that perspective too this is a really difficult decision but performance is also a factor that drives people away so I'm willing to say that they are both equally important and that makes it extremely hard to choose one over the other in my opinion I think we need to go for performance, performance is also a constant struggle with my experience.

And there are days to where I would unfocus out of the session someone does something really stupid in a session I'm not focused into and my whole game ends up crashing because of that. and new users are really confused about why the performance is bad because they're in multiple worlds and they just don't know about it the multi process architecture would help with that and just the ideal of a world freezing up but my dash would be still fully usable to exit the world is invaluable to me

**Reply by Frooxius** [reply]
> IK is something I definitely struggle with and I have a lot of friends who refuse to use Resonite due to this.

Have them fill out the survey if you can! The more data we get the better. But it might also be interesting when it's contextualized as well - e.g. will they need both to be improved before they play? Or would they be fine with current level of performance if IK is improved?

**Reply by ModernBalloonie** [reply]
Will do!

**Comment by Draconas** [comment]
IK would mostly benefit the people who use FBT

however poor performance affects everyone

**Reply by Snerfoil** [reply]
better IK would b nice featurewize, however improving the performance is gonna be healther in the long run. I'm not super worried of having things to get new ppl on board (especially given that often times their choice is based purely on vibe.)

**Comment by PraxThing** [comment]
Some are bothered by the IK, I agree it should be addressed.   But _everyone_ is bothered by the stutters and performance issues.  

It is the one thing _everyone_ has in common:  the experience of asking someone to repeat themselves, or getting disoriented from a bad hitch that pauses rendering and voice.  Especially on join/leave.

I will absolutely increase my support if the performance issues are addressed and the team continues to welcome input like this.

**Comment by lxw404** [comment]
I have actually held off on recommending Resonite to specific individuals who I know are motion sensitive to lag in VR games.  There are several even just minor performance-related issues which can trip up individuals and even perpetuate the false notion that "[Resonite isn't multithreaded](https://www.youtube.com/watch?v=9qGTfV42Beo)" since they seem to lock up the renderer and generally lead to an overall laggy feel:

* [Session freezes when users join](https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1477)
* [Unpacking ProtoFlux freezes all users](https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1482)
* [First touch on colliders causes freeze](https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/419)
* [Opening an inspector on a large slot or a slot with continuous updating bounds freezes all users](https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/2070)

Among many other small issues like this, I'm sure the larger ones mentioned in the post will address a number of initial complaints that I too often hear from new users, which in my opinion should be prioritized above any other feature.  IK would certainly would be exciting and fun addition though, but probably is a lower pain point at the moment comparatively.

**Reply by Frooxius** [reply]
If we work on performance next, it will get significantly better, but please not that preventing every single hiccup and freeze is not really feasible - they will still happen depending on what happens in the session, just less often and to lesser degree.

**Reply by lxw404** [reply]
That is great to hear.  It's certainly completely understandable that not every possible source of lag is preventable, especially given that anyone could spawn tons of objects at once, or just generally not care about optimizing their own creations.  The issues noted above, I believe are things out of users' control and often a source of pain that I hear about constantly, so I would hope some are addressed as part of a performance-oriented update as well sometime.

**Reply by Frooxius** [reply]
Yeah. We definitely want to improve and remove a whole bunch, but I also want to be upfront that there still be some sources of lags and hiccups. Like no matter what we do, I don't think they'll ever completely go away, given the varied nature of the platform.

**Reply by PraxThing** [reply]
@Frooxius Less often and to significantly lesser degrees will vastly affect the overall experience I think.  And in practical terms, it could raise the bar for maximum tolerable concurrent users in a session.  If we have the same experience today with 10 users as we might do in the future with 50 users, that is transformational.

**Comment by PointerOffset** [comment]
While there's lots of reasons to want IK improvements, I think a focus on performance wins out easily as a priority.

The current IK system has problems. However, it is functional and its problems can be worked-around with some advice and community tools. That is *not an ideal or desired situation* but it is, for now, workable.

The same can't be said for performance. There's many things the community can do to improve their experience with IK. There is very little the community can do to improve CPU performance. Some mods and best practices can help, but it's mostly down to FrooxEngine being improved. It's out of the community's hands.

Given the wide scope of frustration poor performance can create, I think improving it greatly outweighs improving IK. Even if the IK is rough for now, it will at least be overall more enjoyable to use Resonite when it performs better. At the very least, it will make the time spent fiddling with the IK more tolerable in the meantime.

**Comment by stawze** [comment]
The low frame rate is one thing, but what I encounter the most (on both Ryzen 5500 and 5800X3D) is frequent freezes and hiccups. I'm used to it happening when a user joins a world or changes their avatar, but it happens seemingly randomly in Resonite, whereas I never noticed this happening in Neos in the many hours I played. I don't know if I just ignored it or got used to it though.

On the other hand IK as been not much of a complaint for me. It would be nice if it could handle weird armatures better, or look a bit more natural, but with a regular humanoid avatar or head-and-hands avatar things work fine right now.

**Comment by hazre** [comment]
How I see it is that people with FBT are always the minority, the loud minority to be exact. Performance would benefit everyone, including desktop users. It would also help with having larger sessions (Clubs might finally become a reality), which leads to more events which will naturally lead to more outsiders checking out Resonite, which I think is the most important thing.

Also FYI this is coming from someone that uses FBT basically everyday, so while I would love a improvement to the IK, I think it's rather selfish when a performance improvement would benefit a lot more people that use the platform on the daily.

**Reply by PraxThing** [reply]
I am also an FBT user and I fully support performance-first.

**Comment by CoffeeKei** [comment]
I personally think performance should be prioritized as it has a larger impact for all users and projects, however I think there are still good questions to be asked.


Question 1: How long does the team estimate each task to take? 

Question 2: Would reworking the IK now require spending additional time later to update it to .NET 8? If this is the case, it may be worthwhile to get .NET 8 in earlier from a resources perspective.

Question 3: What major challenges does the team forsee needing to overcome for each? Any milestones?

**Reply by LexiBasilisk** [reply]
1. "As long as it takes." I don't know, and we'd probably need to start to get into it before being able to make an accurate estimate.

2. Updating runtimes is a lower-level thing, and I believe it wouldn't really directly impact the implementation of the reworked IK one way or another.

3. From the perspective of QC and testing these new features/functionality- "performance" has the potential to be monumentally more difficult compared to IK. E.g. ensuring there are no visible/obvious regressions across all major features and content on the engine. There would be a much wider range of things to touch, depending on which specific performance items we decide to tackle. Reworking the IK would be a lot easier to test because if it is misbehaving, it would be very visually obvious- and it is relatively easy to set up test cases that ensures the new IK behaves as expected.

**Reply by Frooxius** [reply]
1. IK will probably take less time than performance, but they're both big projects, so I would say probably months to be safe. How much exactly will depend and will become more clear once we start working on those.

Performance also has more "pausing points", where we can refocus on other things for a bit and then return to it, because there's a lot more to do overall.

2. IK would work out of the box with .NET 8, but being already on modern .NET and C# will give us better tools to rework it later and make it more optimized from the get go

3. These projects are their own major challenges. Specifics will depend once we focus on these more. 

IK itself we'll need to make sure it works well with all existing avatars and lots of new ones, so there's a lot of fine tuning there.

Performance depends on which way we'll go. The results there are a lot less certain, on what things will help and how much, so there will be some exploratory work and testing. We'll also probably need to cleanup a lot of the codebase so it's more isolated and can communicate over IPC.

**Reply by Amunak** [reply]
> Performance also has more "pausing points", where we can refocus on other things for a bit and then return to it, because there's a lot more to do overall.

Is there a chance that you would decide to omit some large part of the proposed changes (like, say, datatype negotiation for the session protocol) because it would take more time for the overall performance rework, and that would in turn cause issues (more technical debt) for the future?

In other words, if this does take several months, will you feel pressured to release *something* and maybe even pause performance work before all your targets are met to work on other things? It's something to think about.

**Reply by Frooxius** [reply]
@Amunak The overall goal is to do enough things to make performance feel significantly better. Then we can re-evaluate if we want to switch things up for a bit. 

This is also for the major focus, but generally we'd still mix in some other smaller things in the meanwhile, so there will be releases along the way.

**Comment by DrShadowfish** [comment]
I would have to side on Performance, as it stands worlds max out at 20 unless extreme things are done using layering/toggles/active flux culling(even then active culling kills cpu overall for small gpu gains) , cpus and gpus not being utilized to the full potential  .  cant expect everyone to run a X3D or current gen 5+ghrz chips to brute force the few threads the game uses . tbh VRCs main selling point is the fact you can have 60+ ppl in event /club worlds/hangout worlds. if this was possible with reso i can see major hub servers  being a thing promoting more use of public worlds again.

**Comment by BlueCyro** [comment]
As much as I'd like to blow up the IK into a thousand pieces for how janky it can be sometimes, performance issues exacerbate it tenfold.

The number one thing I hear from new users just entering the platform is how surprised they are when the game runs far below what they expect for their spec of computer. Most people who check us out are coming from other platforms where the bar for performance is quite a lot higher.

Obviously this depends on the contents of the world/peoples' avatars, but they're used to a ballpack of somewhere in the high 45's or even 90s of FPS with like 10-20 other people in a beautifully-detailed world.

With 10 people in a moderately-detailed world like the cloud home in Resonite, you're marveled at if you get above 45fps, and that's on high-end processors like AMD's 7800x3D with a lot of cache to work with. Anything less than cutting-edge hardware has you sitting maybe somewhere in the mid-30s or even 'teens in that scenario. That's even ignoring the gut-wrenching hitches that happen when a user spawns in.

Performance. Is. Paramount. If we want users to be able to live comfortably in an extremely connected and cooperative platform, having the comfort of a nicely-running simulation is essential for things like get-togethers, build jams, outright parties, or even virtual conventions. Being able to effectively use Resonite for its intended purpose without worrying about the next garbage collector sweep or seconds-long pause from users joining I think will be the tipping point in *really* getting people interested and excited about our little section of the internet we call home.


Perhaps that's a bit of a blunt review of our current situation, but I feel it's an accurate one, and I feel it's always been the kryptonite to our vision: A truly massive, collaborative digital universe.

**Reply by BlueCyro** [reply]
As an addendum to this, I've been experimenting with running the Headless server on .NET 8 and have seen very, very promising performance improvements just from the runtime switch.

Bepu (our physics engine) in particular runs significantly faster, but that was always expected.

Session performance has been nothing but stellar. While I don't have a direct comparison for mono; 13-20 players un-culled, on a Ryzen 5800x, I saw an uncapped server tick rate of over **240fps**. This is obviously without any graphical output, but was with players coming and going, spawning vehicles, spawning protoflux, swapping avatars, and just generally being a normal active session.

I've also witnessed usage of .NET 8 for various (sometimes quite large) party sessions with over 30 people and the server **never** fell below its set 30fps or 60fps tickrate for more than a single update.

Take that with maybe more than a pinch of salt since it's just the headless server, but the particular horizon of a runtime upgrade to .NET 8 looks very promising.

**Comment by TisFoolish** [comment]
Bug report with the form: The "Other" form under the question "What is a major issue that currently makes it hardest for you to enjoy Resonite?" is unable to be toggled in both Chrome and Firefox for me

**Reply by Nytra** [reply]
I think once you start typing in the box it enables it

**Reply by Frooxius** [reply]
You just need to type your answer in the box. It seems it's just how NextCloud "Other" option works unfortunately.

**Reply by TisFoolish** [reply]
Ahh, that is unfortunate as that is not intuitive for me and probably many others

**Reply by koyae** [reply]
Yeah, I got really confused by that as well. Got halfway done filling it out on an old Android phone before I thought it was breaking due to me, then figured out that weird behavior after getting onto desktop and finding it was still "broken," and then guessing I might need to type something first.

In this or future surveys done through NextCloud, you might consider just noting that behavior in parenthesis on questions that offer an "Other" option. That way folks won't be tripped up.

**Comment by WinterCaverns** [comment]
If we get performance done first, it'll help build the player base. It will also make other changes such as IK more impactful.

Performance can take weeks if not months to implement. With that said there can be hurdles. 

IK would be quicker to implement overall and then focus on the major part.

**Comment by ErrorJan** [comment]
Probably performance should be prioritized first. Every time I play a different VR Game and then come back to Resonite I immediately feel the difference in performance and that is quite frustrating, since this is a platform I really enjoy using.
Would probably also make new users stay more, since performance is the first thing you notice. IK, while it feels jank, can be overlooked here and there. You will always feel when the performance drops though.

**Comment by lill-la** [comment]
Please do not use css `user-select: none` in survey page.
It makes it difficult to get help from translation software.

**Reply by ProbablePrime** [reply]
We're using a stock version of Next-cloud's form application for this survey. 

We had little time to add our own branding or design but will keep this in mind for future surveys.

Why NextCloud btw? DATA PRIVACY/PROTECTION BLARGH

**Reply by lill-la** [reply]
We already have friends who have given up on answering the survey because they cannot get help from translation assistance software.
It is also important to remember that there are people for whom non-English language support is more important than performance or IK.

**Reply by Frooxius** [reply]
The survey will be open for 2 weeks, that should be enough time to help with translation of it so they can fill it in.

I'm not sure how much we can control css and the code, given it's NextCloud. Are there any other things we could do here? Would translating it by copying & pasting the text work? 

We could perhaps number the options and have a translated version available separately?

For the non-English language support - Resonite already supports non-English languages, we have a number of different locales. Are there any major things missing in this regard?

**Reply by ProbablePrime** [reply]
I'll see what I can do.

**Reply by lill-la** [reply]
> The survey will be open for 2 weeks, that should be enough time to help with translation of it so they can fill it in.

Regardless of the time period, some people give up answering when translation is difficult.

> For the non-English language support - Resonite already supports non-English languages, we have a number of different locales. Are there any major things missing in this regard?

I generally see no problem with it.

**Reply by lill-la** [reply]
I talked to some friends and apparently there is a different problem than I thought.

I thought it was because the Q&A texts were not selectable so they could not be copied and pasted, making it difficult to run through translation software, but apparently people use their browser's translation tools more often than not.
And apparently the survey page doesn't work with Chrome or Edge translation tools for some reason.

**Reply by lill-la** [reply]
```
<html class="ng-csp" data-placeholder-focus="false" lang="ja" data-locale="ja" translate="no" >
```
Apparently, the problem is that "translate="no" is attached.
Without it, the problem would probably be solved.
(Personally, I would be happier without "user-select: none" as well.)

**Reply by lill-la** [reply]
> I'm not sure how much we can control css and the code, given it's NextCloud. Are there any other things we could do here? Would translating it by copying & pasting the text work?

> We could perhaps number the options and have a translated version available separately?

Perhaps another resource would be available, but people who cannot read English would not be able to get to it.
So far I have been able to support my friends in the communities within my reach, but it will be more of a problem if the Resonite population grows.

**Reply by ProbablePrime** [reply]
I have direct access to the bare metal machine that runs this survey so if its just removing a property from the HTML I will go aggressively do that!

**Reply by ProbablePrime** [reply]
I have manually edited the PHP code on the machine to remove translate=no.

Can you try again? 

Oh you can read about this problem here: https://github.com/nextcloud/server/commit/087deb9f1b45b9d9227cf5e82a18b21c6f788ff6

I'll be following up with the next cloud devs to... figure out why.

**Reply by ProbablePrime** [reply]
More information here: https://github.com/nextcloud/server/issues/34857

Verified: 
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/8791132/65827d2b-c28e-4758-8362-199d4173ab48)

**Reply by lill-la** [reply]
It's working fine.
Thank you very much.

**Comment by GooseVali** [comment]
The biggest thing personally that bothers me among these issues is the momentary freezes, especially those that intervene with speech audio and talking to people. It adds extra hurdles to socializing with people. For me that's a bigger issue than overall lower average framerates. It's frustrating to have your communication mangled by someone joining or some other freezing issue.

There's also the other stutters and hitches from various sources. It would be nice if loading heavy things could be done in the background without bringing the world to a halt momentarily. Would probably be preferable to not see a newly spawned item for 3 seconds instead of the world freezing for 1 second and the item then appearing.

**Reply by Frooxius** [reply]
Working on performance would definitely improve a lot of the freezes and remove some of them completely.

Just to give more context though, loading of things actually is done heavily in the background!

However with current runtime, that actually causes some things to freeze due to heavy memory allocations, due to poor garbage collector, which will freeze the entire app at times to free up memory. The upgrade to .NET 8 would improve this significantly, since it includes significantly more efficient garbage collector (among other things).

**Comment by LakesideMiners** [comment]
preformance, its the  main thing that (along with UI) that holds back Resonite imo. IK will also have more room to use resources if the main stuff gets optimized.

**Comment by CheesecakeCG** [comment]
Performance has been the only thing stopping me from using Resonite and the old platform since I'd first tried it... but my biggest worry that if there is a big performance update, people in similar situation as me will rush in, discover the current IK system and be put off again. So, it could make sense to improve the IK system and maybe some more UX polish before - at the very least - advertising any huge performance improvements. Any improvements sooner are obviously still welcome, but too people knowing Resonite suddenly runs much better too soon might be opening Pandora's Box. :p

**Reply by Frooxius** [reply]
Once we do performance improvements, we will likely announce it as soon as it's done (plus there will be devlogs and other stuff).

We generally develop pretty openly, rather than being quiet, working on several big things, before releasing them all, so that wouldn't really fit well with how we work unfortunately.

There's always the risk that other things will turn people away - but I don't think that ever really goes fully away. It's one of the reasons we're asking people in the survey if they would need both to be improved to play, or if just one is sufficient - if more people would play more even if just one of these gets addressed, that's a pretty good indicator that it's worth it.

**Comment by GriffinPhillis** [comment]
Definitely am looking forward to _both_ of these updates, aah!! 💚
As for what should come first, as a lot of others here seem to be leaning towards, I too think that performance updates should take priority.
I know many friends who want to play Resonite but have to choose against it solely because their PCs can't handle it well enough. I don't get to play it as much as I'd like to, either, for the same reason! Performance boosts, though? I think that would bring more people and activity in. <3

**Comment by gameboycjp** [comment]
I wish there was an option for either would make me play more. While performance seems to be the community consensus, and mine, for what's more important, I would be perfectly excited for either.

**Comment by modimobeikete** [comment]
In my opinion, the performance is a must to fix. I know that most of Chinese player said they can't stand for the client freeze when they entered another world. 
I guess this is the only way to expand this community. Otherwise they are not likely to join even there are some more features which is better than Vrc.

**Reply by modimobeikete** [reply]
Besides, some of them said the official tutorial was not comprehensive and clear. They think Resonite Community Tutorial World is much more clear.

**Comment by DominJgon** [comment]
Performance seems to be the main issue for most players at the moment, especially when new users join and are met with constant hiccups. As for the lack of IKs, for me at least, they are sufficient at this stage of development. Any special cases for IK can be temporarily user-created via protoflux if someone needs it.

Many tools are evolving into more and more complicated abstractions, which could also use a bit of any performance boost that can be made.

There's also the issue of increased security, which is very important for projects that allow code execution and communication outside the application, such as http requests.

**Comment by An12854** [comment]
PLEASE focus on performance, i have an R5 4600H which is the mobile equivalent of the R5 2600 and my frametime looks like a heartbeat graph with an RTX 2060 on the home world in desktop mode
![imagem](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/50544887/5aa55a95-1e6e-48b7-b681-2557a4bcd7d3)

6 CPU cores are the most common config on steam and 99% of people trying to play this game (including me) aren't going to buy an X3D CPU for this specific reason

**Comment by SizemoreSystem** [comment]
I'd like to say that I think that for people who have a bit less money focusing on performance first would definitely help a lot with accessibility, since full body tracking costs a lot of money and big fancy computers cost a lot of money.

**Reply by ProbablePrime** [reply]
Remember that FBT isn't the only reason we'd be doing the IK update.

**Reply by SizemoreSystem** [reply]
What else would it affect?

**Reply by ohzee00** [reply]
I would suggest reading the issue here: https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/612

**Reply by ProbablePrime** [reply]
And read the survey itself:

> What will IK rework do?
> Most significantly, it will improve how full body avatars look and feel
> Full body tracking will behave better - eliminating neck scrunch, better hips response and so on
> Add new capabilities - shoulder tracking, possibly more than 2 arms & legs, maybe quadruped support
> Likely improve avatar support - supporting various skeletons and rigs (like digitigrade legs)

Only one of these items is strictly FBT support.
For example, Digitgrade Legs + Quadrupeds is huge for some users here.

Remember that IK affects how your avatar moves and feels when moving, better IK here can potentially make you feel more at home in your avatar and that's important too.

**Reply by Electronus** [reply]
god quadruped support would be so nice. 

not only would it make "feral" avatars easier but Taur ones as well!

**Reply by SizemoreSystem** [reply]
> And read the survey itself:
> 
> > What will IK rework do?
> > Most significantly, it will improve how full body avatars look and feel
> > Full body tracking will behave better - eliminating neck scrunch, better hips response and so on
> > Add new capabilities - shoulder tracking, possibly more than 2 arms & legs, maybe quadruped support
> > Likely improve avatar support - supporting various skeletons and rigs (like digitigrade legs)
> 
> Only one of these items is strictly FBT support.
> For example, Digitgrade Legs + Quadrupeds is huge for some users here.
> 
> Remember that IK affects how your avatar moves and feels when moving, better IK here can potentially make you feel more at home in your avatar and that's important too.

I saw that, but didn't think much of the last line. I still think that for accessibility, performance should be worked on first. That's my personal opinion though.

**Comment by Astral-J** [comment]
Out of curiosity, for the IK rework, would that fix hips being able to pull your head out of position? i.e. instead of the hip tracking having the main priority would it be changed so that the head would have the greatest priority?

**Reply by Frooxius** [reply]
Yes that will be fixed with IK rework.

It won't necessarily be with hips or head having main priority - that might not even be a concept there, depending on how exactly it would end up working.

**Reply by Lich-King-Ethan** [reply]
this might be weird, but have you considered doing mixed tracking or an optional toggle for a built in forward kinematics system? Slime is great, but a system directly integrated into Resonite would be something interesting to see, or a plugin of some sort for IMU based systems to communicate directly? i understand it would be a lot more work, but the potential reductions in latency/programs to run could also be a performance boost of its own! :D

**Reply by Frooxius** [reply]
It'd be best to make feature request for something like that. It doesn't sound like something we considered, but I'm having trouble understanding what you're actually asking for. 

It sounds more like a driver thing? You can't really drive avatars with forward kinematics like that, that wouldn't result in good behavior, as your limbs and such wouldn't end up where your physical ones are most of the time.

This wouldn't really provide any performance boosts in any case.

**Comment by Lich-King-Ethan** [comment]
So as someone who loves Resonite's current feature set and sheer ironclad versatility coming from like, a smidge on VRChat (less than 500 hours across multiple platforms options) and a few hours at most on Resonite trying out the tutorial world and then getting my blender edited Vroid avatar imported~ish, rocking a 3070Ti, 12600KF, 32GB of DDR4 RAM @3600MHz and a Quest 3 with touch plus controllers and hand+body tracking enabled, running wireless through Virtual Desktop at High settings, usually 90FPS mode, with SSW set to auto and SSGR enabled, PlayStation Move Service Ex 6 point full body optical tracking, I can say it could take the performance improvements before the full body personally.

 Reason being, i believe if we also want to increase potential player base counts as well, having a big performance update would be a great thing both to advertise, but also for the longtime users to boost frame rates or stability, while also hopefully lowering base requirements in the process. I also understand i cannot and should not expect like, 50% higher frame rates or anything absolutely nutty, but I'm not gonna lie, even the tutorial ramps my system up pretty good!

 But i do feel that higher performance across the board, lowering entry requirements, is a great way to steer Resonite for a little while i think! :)

**Reply by Lich-King-Ethan** [reply]
i also want to clarify, i never got my avatar imported and set up enough to test my FBT solution with it yet, but we'll get there one day, just gotta get the motivation! lolll

**Reply by Frooxius** [reply]
Thank you for sharing, but remember to fill out the survey too if you haven't!

**Comment by TisFoolish** [comment]
It's pretty obvious that the community is leaning towards performance over fancy IK, My question is, will we be getting the performance updates piecemeal or all at once? EG like the switch to .NET 8

**Reply by Frooxius** [reply]
We'll do them as piece meal as possible, but some are very big chunks of work that will have a lot of work that needs to be done before it's functional.

**Reply by Readun** [reply]
Would it also include things like Probableprime found out about the Axis aligner and other aligner comonents updating with every frame?
I imagine doing the easier goals first before approaching the big rewrites would maybe already help a little?

**Reply by ProbablePrime** [reply]
That axis aligner one is just an example.

We're going to poke around and get in deep in all the crevices and find out what the major issues are and fix those. Due to this we don't know what will fit into the update but we do know that it will focus on performance.

It of course depends on the individual issue but please try to imagine the performance update as a very large area to work on with many different avenues for exploring. Its as much a research project as it is an engineering one. Due to the research element, its not exactly possible to state what is and isn't in the update in some cases.

**Reply by Frooxius** [reply]
The focus is on global performance improvements, so there is no decision made on what specifically will be improved and what will not.

Part of work on performance is heavy profiling. If we find that particular component is very common source of performance issues, we might focus on it. 

But if it's more niche, then we might not pay attention to it specifically.

**Comment by koyae** [comment]
I've yet to review everything here, so forgive me if any of these things have been asked previously! Feel free to link stuff and/or quote from other posts.

1. *To what extent can better IK solutions simply be **bought?*** Or — better yet — *bought and then customized?* There are many IK systems out there (some of which are for sale), whereas my experience with performance optimization is that it is rarely is it possible to simply spend (or search through open-source libraries and integrate) one's way out of "the application (or some aspect of it) is slow," other than for very common bottlenecks which drive large demand for solutions.
2. *To what extent are Resonite users locked into the IK system?* OpenVR/OpenXR are inherently trick-able, and systems like OSC ought to allow users to run practically any pose-estimator they desire locally, then pipe the results over the network. Obviously, telling users to DIY something has costs (e.g. network overhead, places a bit more burden of technical knowhow on users), but it seems more tractable than telling folks to roll their own performant version of Resonite. Does that sound about right?

**Reply by ProbablePrime** [reply]
1. We don't intend to buy or rely on other IK Solutions for the IK update. Froox will be writing a 100% custom in-house IK
2. It is possible in the future you might be able to write your own, this doesn't really affect the timeline of our in-house system though. The new system will likely be more customizable which may reduce the need to write your own. As a secondary, if you feel the need to write your own right now please let us know why. It might be something that we can fix.

**Reply by Frooxius** [reply]
1. The current IK solution has been licensed and customized. This has reached its limits, which is where we are with the current IK in Resonite. To make things work better, we need an in-house one that's custom tailored to Resonite's needs.

2. You can roll your own IK system if you really want to with existing tools already. A few users have avatars which do this actually. However this is extremely user intensive and doesn't solve any of the problems most users are having and places a pretty high burden on users to get their avatar working.

Pose estimator is insufficient - that tends to only estimate where your body is, but you still need to solve the pose of avatar bones to match that.

**Reply by koyae** [reply]
@ProbablePrime @Frooxius My confidence in this project has risen already cuz hot dang y'all're fast. I'm off to submit my survey answers now. Thank you for the clarifications and for keeping the platform NFT-free :D

**Comment by Mr-CostlessClaw** [comment]
Imma put my two cents in.
I am also voting performance. Hopefully this have a noticeable improvement while waiting on the new graphics engine. 
Though one idea I had was if there is a fast/quick IK hotpatch that could be rolled out.
But I am still on for performance since it affects everyone and I already know of people who can't play due to world shudders crach the game.

**Reply by Frooxius** [reply]
There's not a fast/quick IK hotpatch that could be done.

We kept applying a number of hotpatches to it and it has reached its limits, which requires a more substantial overhaul.

**Reply by Mr-CostlessClaw** [reply]
It was a suggestion, but I am looking at performance mainly. Use a steamdeck on the side and this make the game playable for more people.

**Reply by Mr-CostlessClaw** [reply]
And I really wanna swe what this game would do to my quest pro~ ^^

**Comment by ProfessorDey** [comment]
My primary interest as a linux user is very definitely performance. As it stands on a Ryzen 7 3700X and a RX6950XT I can still only play the default home in desktop mode at 25-45 fps, and that barely changes whether at 1080p or 4k. My biggest concerns are thus culling and asset loading times.

Just starting the default home takes over a minute of halting progress as it loads each texture and audio clip and I regularly need to enable swap to get past that point as it completely consumes my 16GB of RAM. I've had to set up one world to use Value overrides to only enable parts of the map when I'm contacting a collider just to reduce the chugging to managable levels.

So the sooner we can get a more robust and efficient data model the better in my book. Gotta love the struggle for supremecy that is Structure of Arrays (data oriented) vs Array of Structures (object oriented).

**Comment by ProbablePrime** [comment]
Should this now be closed? The survey is over!

**Reply by LexiBasilisk** [reply]
Yeah. This thread was primarily created to allow users a space to ask questions to better inform their choices while voting- as the polling has since wrapped- this can be closed.
